atlasdb-commons/src/main/java/com/palantir/common/proxy/ReplaceIfExceptionMatchingProxy.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.common.proxy;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class ReplaceIfExceptionMatchingProxy<T> implements InvocationHandler {
+    private static final Logger log = LoggerFactory.getLogger(ReplaceIfExceptionMatchingProxy.class);
+
+    private final Supplier<T> delegateFactory;
+    private final Predicate<Throwable> shouldReplace;
+    private volatile T delegate;
+
+    private ReplaceIfExceptionMatchingProxy(Supplier<T> delegateFactory, Predicate<Throwable> shouldReplace) {
+        this.delegateFactory = delegateFactory;
+        this.delegate = delegateFactory.get();
+        this.shouldReplace = shouldReplace;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        try {
+            return method.invoke(delegate, args);
```

Comment:
I think we need to compare the delegate that we actually invoked the method on, not the value after an exception is caught. Concurrent failures could cause us to roll several times.

Reply:
This is handled outside of this class. The supplier given here is memoized.

Reply:
I don't understand, in that case wouldn't we always replace the delegate with itself? I just rolled out of bed though, and might not be reading thoroughly as I should...

I think we need something like:

```java
T localDelegate = delegate;
try {
  return method.invoke(localDelegate, args);
} catch (ITE e) {
  Throwable cause = e.getCause();
  replaceIfNecessary(cause, localDelegate);
  throw cause;
}
```
Then update `replaceIfNecessary`
```java
   private synchronized void replaceIfNecessary(Throwable thrown, T invocationTarget) {
        if (shouldReplace.test(thrown)) {
            // cas on fail
            if (invocationTarget != delegate) {
                // already replaced
                return;
            }
            T replacement = delegateFactory.get();
            if (delegate != replacement) {
                log.info("Replacing underlying proxy due to thrown exception", thrown);
                delegate = delegateFactory.get();
            }
        }
    }
```

Reply:
Ok so reasoning through this:

The supplier passed here is:

```
Suppliers.memoizeWithExpiration(supplier::get, 20, TimeUnit.MINUTES)
```

So when we hit the bug, all requests will enter the synchronized block and one by one, check they hit the bug, then call the factory. Only the first one will actually refresh the client (and the supplier will memoize it for the next 20 minutes). Next one will call the factory again and get the same one and so not replace it.

The only thing I'm wonder is why is the delegateFactory called twice

Reply:
Talked offline: the proposed code from @carterkozak is to make sure that if we ever switch the supplier passed to here, we don't shoot ourselves in the foot. I will keep as is, as we hope to remove this shortly.